src/Scale.hs

```diff
@@ -1,99 +1,199 @@
-scaleToIntervals SDorianb2 =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 2
-scaleToIntervals SLydianAug =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 3
-scaleToIntervals SLydianDom =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 4
-scaleToIntervals SMixolydianb6 =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 5
-scaleToIntervals SLocrianSharp2 =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 6
-scaleToIntervals SAltered =
-  nthDegreeIntervals (scaleToIntervals SMelodicMinor) 7
-scaleToIntervals SHarmonicMinor =
-  listIntervals
+baseModeIntervals LydianAug =
+  nthDegreeIntervals (baseModeIntervals MelodicMinor) 3
+baseModeIntervals LydianDom =
+  nthDegreeIntervals (baseModeIntervals MelodicMinor) 4
+baseModeIntervals Altered =
+  nthDegreeIntervals (baseModeIntervals MelodicMinor) 7
+baseModeIntervals HarmonicMinor =
+  zipToIntervalSet
   [Perfect, Major, Minor, Perfect, Perfect, Minor, Major] [1..7]
-scaleToIntervals SLocrianNat6 =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 2
-scaleToIntervals SIonianSharp5 =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 3
-scaleToIntervals SDorianSharp4 =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 4
-scaleToIntervals SPhrygianDom =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 5
-scaleToIntervals SLydianSharp2 =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 6
-scaleToIntervals SSuperLocrianbb7 =
-  nthDegreeIntervals (scaleToIntervals SHarmonicMinor) 7
+baseModeIntervals PhrygianDom =
+  nthDegreeIntervals (baseModeIntervals HarmonicMinor) 5
+baseModeIntervals DoubleHarmonicMinor =
+  zipToIntervalSet
+  [Perfect, Major, Minor, Augmented 1, Perfect, Minor, Major] [1..7]
+baseModeIntervals HarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Major, Major, Perfect, Perfect, Minor, Major] [1..7]
+baseModeIntervals DoubleHarmonicMajor =
+  zipToIntervalSet
+  [Perfect, Minor, Major, Perfect, Perfect, Minor, Major] [1..7]
+
+
+modeToIntervals :: Mode -> Set Interval
+modeToIntervals (Mode baseMode exts) = foldr extIntervals (baseModeIntervals baseMode) exts
+  where 
+    extIntervals :: ScaleExt -> Set Interval -> Set Interval
+    extIntervals ext intSet = insert (oldInt <+> (impliedShift $ acc ext)) (delete oldInt intSet)
+      where
+        -- TODO: If there isn't only one interval of a certain degree, the mode is
+        -- ambiguously constructed and we should give a warning.
+        oldInt = elemAt 0 (S.filter (\a -> getSize a == deg ext) intSet)    
+
+
+scaleToNotes :: Scale -> Set Root
+scaleToNotes (Scale root mode) = mapMonotonic (flip jumpIntervalFromNote root) (modeToIntervals mode)
+
+
+modalDistance :: Set Interval -> Set Interval -> Int
+modalDistance mode1 mode2 = sum $ intDistance <$> zip (toAscList mode1) (toAscList mode2)
+  where 
+    intDistance :: (Interval, Interval) -> Int
+    intDistance (i1, i2) = abs $ fromJust $ intervalToDistance (i1 |-| i2) 
```

Comment:
All I could find was Data.Sized which seems like a vector thing rather than a natural numbers thing...am I missing something or is it a diff type class?

Reply:
Ah, sorry - I meant that it's currently possible for the `intervalToDistance` function to return a negative number, when it should always be returning a positive value (unless your notion of distance was intended to have directionality, but I'm not sure this consistently happens in the implementation.)